api/STRUCTURE.md

```diff
@@ -0,0 +1,83 @@
+
+TODO: Why do we do this this way?
+
+Services / ServiceImpls are generally named after a feature or functional area, like 'UserService', 'CohortService', or 'MonitoringService'.
+
+ServiceImpls can call into other services, but we try to keep the number of dependencies down. We try to avoid having ServiceImpls call into a large number of DAOs, especially DAOs not directly related to the functional area for which the ServiceImpl is responsible. For instance, the `UserServiceImpl` calls into the `UserDao`, the `AdminActionHistoryDao`
+
+ServiceImpls should handle only application logic. They should not need explicit knowledge of the database, the REST API, or the UI.
+
+## Controllers
+
+Controllers take Request objects from the API, call into the Service layer, and return Response objects back to the API. Controllers implement autogenerated [APIControllers](#apicontrollers).
+
+Controller methods should generally be small and should not contain any actual logic. The most common conditional in the Controller layer should be checking a feature flag. Most of the work done by Controllers should be munging data provided by the API or a Service.
+
+### Mappers
+
+There are four types of models in the system - [API models](#swagger), [DB models](#db-models), and [API Request and Response](#requests-and-responses) models. Mappers are utility classes that own methods that convert between these types of models.
+
+## API
+
+The actual REST API is defined in a series of YAML files according to the [Swagger OpenAPI spec](https://swagger.io/specification/). Swagger takes these YAML files, merges them, and generates all the classes defined by the YAMLs. The command to do so is `./project.rb compile-generated-java`.
+
+One non-Swagger-standard thing about these YAML files is that the `tag` param is used to indicate which Controller / APIController an endpoint should be associated with.
+
+Each REST endpoint is defined under `PATHS` in the YAML file.
+
+### API Models
+
+API models are defined under `DEFINITIONS` in the YAML file. Swagger autogenerates classes from these. Both API models and DB models are currently used in the Controller and Service layer. There is not currently consensus on whether the standard should switch to only using one or the other.
+
+TODO: is the latter sentence true?
+
+### Requests and Responses
+
+Swagger interprets the data sent to a given endpoint according to the Request model definition associated with that endpoint and throws an error if the data is improperly formatted. Likewise, Swagger validates that the data being returned from a given endpoint conforms to the Response model definition associated with that endpoint and throws an error if it does not.
+
+Technically, Requests and Responses are just more API model classes, similarly autogenerated by Swagger, but they are only used when communicating directly with the API layer. API models that represent business logic concepts are more freely used throughout the system.
```

Comment:
The missing bit here is a business-logic-level class. I think a major answer here will be Immutables.org objects, which let you define ad hoc value objects with builders that aren't subject to mutation.

Reply:
This document is about the current state of the system.

Reply:
🤷‍♂ I'm not sure I'd get into it then. It's kind of a detour.